[WDR] complete overhaul after relaunch of the site (also fixes #8562)

Boris-de opened the pull request:
The WDR relaunched their site on 2016-02-23 which not only changed the
URL-schema completely but also the layout of their pages.

Apparently the whole "mediathek" now runs on the wdr-domain, so no
separate URL for funkhauseuropa anymore.
There seems to be no explicit handling of video-sizes on the page or in
the URLs anymore. There seems to be only one size for HTML5, but still
several sizes for flash. The new extractor adds all to the list of formats.

There is no metadata for the HTML5-stream, so that the best flash-stream
will always be considered as the "best" format. At least in my tests
this seemed to be true anyway.

As this is my first extractor please let my know if I did anything wrong or
missed something.
I was not sure about the tests, currently they are of course working, but
because of the weird laws in Germany I know that they will be deleted
soonish, so I already added the skips to avoid that they'll break the build
at the time that happens.

This also fixes #8562 by just adding the current show URL to the
WDR-extractor, which works fine.

EDIT (2016-03-19): as #8842 was closed as a duplicate of this PR, I added
the text from there, so that there is a complete description on this changes
without having to look into the commit messages.

FelixSchwarz commented:
works for me

z1atk0 commented:
I'm not sure if this site-overhaul-fix is already in the official version or not due to the above conflict, but WDR does not work at all for me in the latest version, it doesn't find a single video:

[zlatko@warrior tmp]$ ./youtube-dl -v "http://www1.wdr.de/mediathek/video/sendungen/wdr-dok/video--kilometer-heimweh---ein-jahr-in-china-folge--106.html"
[debug] System config: [u'--prefer-free-formats']
[debug] User config: []
[debug] Command-line args: [u'-v', u'http://www1.wdr.de/mediathek/video/sendungen/wdr-dok/video--kilometer-heimweh---ein-jahr-in-china-folge--106.html']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2016.04.19
[debug] Python version 2.7.5 - Linux-3.10.0-327.13.1.el7.x86_64-x86_64-with-centos-7.2.1511-Core
[debug] exe versions: none
[debug] Proxy map: {}
[WDR] mediathek/video/sendungen/wdr-dok/video--kilometer-heimweh---ein-jahr-in-china-folge--106: Downloading webpage
[download] Downloading playlist: mediathek/video/sendungen/wdr-dok/video--kilometer-heimweh---ein-jahr-in-china-folge--106
[WDR] playlist mediathek/video/sendungen/wdr-dok/video--kilometer-heimweh---ein-jahr-in-china-folge--106: Collected 0 video ids (downloading 0 of them)
[download] Finished downloading playlist: mediathek/video/sendungen/wdr-dok/video--kilometer-heimweh---ein-jahr-in-china-folge--106

Same problem with current version (2016.04.19 as of today) with both Python-2.6.2 (Slackware 13) and Python-2.7.5 (CentOS 7.2.1511).

Boris-de commented:
@z1atk0 This is not in the official version yet. There were two commits that changed the regexes a week or two after I opened the PR which cause the conflict. Thanks for noticing me about this, I wasn't aware of it. I just rebased my changes on top of the current master, so that the conflict should be resolved for now.

I also tested your example-URL, it also works with this change.

z1atk0 commented:
@Boris-de Great, thanks! :-)

FelixSchwarz commented:
btw: What's the process to get this merged? It works for me and it's a bit annoying having to use this fork instead of just the main version.

z1atk0 commented:
What is this Travis CI thing that keeps failing? :-/ Looks like I'll have to get the source code and apply your two commits as diffs then for the time being, I guess ... oh well.

FelixSchwarz commented:
Travis a web services which executes youtube-dl's self-tests to ensure that all other functionality still works after adding a new feature. Usually new code won't be committed unless all tests still pass. However youtube-dl's tests are a bit special because they try to download pages from real web servers (so they are not isolated) and changes on some of these servers will cause travis failures. I think the failing tests have nothing to do with the new code from this pull request so the code should be merged anyhow.

z1atk0 commented:
Ah, got it, thanks. :-)

yan12125 commented on youtube_dl/extractor/wdr.py:
```diff
@@ -11,204 +10,168 @@
+        webpage = self._download_webpage(url, display_id)
 
-        webpage = self._download_webpage(url, page_id)
+        js_url = self._search_regex(self._JS_URL_REGEX, webpage, 'js_url', default=None)
```
I suggest extracting from ```mediaLink``` element instead of matching the JS URL as it may change in the future. For example:
```Python
        media_link_obj = self._parse_json(self._html_search_regex(
            r'class="mediaLink\b[^"]*"[^>]+data-extension="([^"]+)"', webpage, 'media link'),
            display_id, transform_source=js_to_json)
        js_url = media_link_obj['mediaObj']['url']
```

yan12125 commented on youtube_dl/extractor/wdr.py:
```diff
@@ -11,204 +10,168 @@
-        flashvars = compat_parse_qs(self._html_search_regex(
-            r'<param name="flashvars" value="([^"]+)"', webpage, 'flashvars'))
+        js_data = self._download_webpage(js_url, 'metadata')
+        json_data = self._search_regex(r'\(({.*})\)', js_data, 'json')
```
Use ```_download_json``` with ```strip_jsonp```. Need to change ```strip_jsonp``` so that the dollar sign is recognized.

yan12125 commented on youtube_dl/extractor/wdr.py:
```diff
@@ -11,204 +10,168 @@
+                if tag_name in metadata_media_alt:
+                    formats.append({
+                        'url': metadata_media_alt[tag_name]
+                    })
```
In http://www1.wdr.de/mediathek/video/sendungen/lokalzeit-bergisches-land/video-lokalzeit-bergisches-land-vom--236.html (from #9517), the ```alt``` URL is a M3U8 manifest.

yan12125 commented:
Always use single quotes for strings.

Boris-de commented:
@yan12125 thanks for all the input. I just made a change for every issue you mentioned.

Additionally I switched wdrmaus.de completely to the new extractor because the other video-pages there now also use their new player. I also fixed a bug in the parsing of playlists that I introduced in the original commit, it seems that there has been a change in the layout and number of available items.

I've pushed the fixes a separate changes for now, because I think it might be easier to review. Just let me know if you prefer a squashed version of the changes or if you have any other comments.

Shuro commented:
I've compiled Boris version for testing, but some WDR-videos seem not to work.  :disappointed: 
> ./youtube-dl http://www1.wdr.de/mediathek/video/sendungen/lokalzeit-ostwestfalen-lippe/video-lokalzeit-owl-738.html

phihag commented:
This is most definitely better than the current version; which failed for many URLs. Merged for now, we'll do future changes in tree.

s-timm commented:
Sorry to add to this thread with a newbie question but I don't exactly follow what's going on: 
are those fixes already included in the "2016.06.03" version or will they be available in a later version? Because I just wanted to download a children video for my son from
http://www.wdrmaus.de/elefantenseite/#/lieder_metcalf_das_wilde_tier
and it doesn't work. :-(

yan12125 commented:
@s-timm These changes will be included in the next version.

Boris-de commented:
@Shuro this URL works for me with the current master. Do you have an error message you can provide? Are you using any special settings?

@s-timm sorry, but the new extractor doesn't work for the "Elefantenseite", and neither did the old extractor as far as I can see (see issue #5731). The site is completely different to the usual video sites in the WDR mediathek or the wdrmaus site. I does seem easy to support though, I can't promise anything but I'll try to create an extractor.
